File: No-Python/src/triage.js
export function classifyIssue(issue, comments = []) {
  const labels = (issue.labels || []).map((label) => label.name?.toLowerCase() || "");
  const text = `${issue.title || ""}\n${issue.body || ""}\n${labels.join(" ")}`.toLowerCase();

  const intent = issueIntent(text, labels);
  const severity = scoreSeverity(text, labels, comments);

  return {
    intent,
    severity: severity.label,
    score: severity.score,
    rationale: severity.rationale,
    recommendedAction: severity.score >= 70 ? "Open an agent PR for review." : "Prepare a focused fix if context is clear.",
  };
}

function issueIntent(text, labels) {
  if (hasAny(text, ["documentation", "docs", "readme", "typo", "example text"])) {
    return { kind: "docs", confidence: 0.78, rationale: "Issue appears documentation-oriented." };
  }
  if (hasAny(text, ["feature", "enhancement", "add support", "new option", "request"])) {
    return { kind: "enhancement", confidence: 0.72, rationale: "Issue asks for new or expanded behavior." };
  }
  if (hasAny(text, ["bug", "error", "fails", "failure", "crash", "regression", "not recognized", "incorrect"])) {
    return { kind: "bug", confidence: 0.84, rationale: "Issue describes broken or incorrect behavior." };
  }
  if (labels.some((label) => ["bug", "defect"].includes(label))) {
    return { kind: "bug", confidence: 0.8, rationale: "Issue labels indicate a bug." };
  }
  return { kind: "unknown", confidence: 0.45, rationale: "Issue intent is ambiguous." };
}

function scoreSeverity(text, labels, comments) {
  let score = 25;
  const rationale = [];

  if (hasAny(text, ["security", "vulnerability", "data loss", "production down"])) {
    score += 45;
    rationale.push("High-impact language appears in the issue.");
  }
  if (hasAny(text, ["crash", "regression", "broken", "fails", "failure"])) {
    score += 25;
    rationale.push("Issue describes broken behavior.");
  }
  if (labels.some((label) => ["p0", "critical", "urgent"].includes(label))) {
    score += 35;
    rationale.push("Labels indicate urgent priority.");
  }
  if (labels.some((label) => ["bug", "defect"].includes(label))) {
    score += 15;
    rationale.push("Labels indicate a bug.");
  }
  if (comments.length > 3) {
    score += 8;
    rationale.push("Issue has active discussion.");
  }

  score = Math.max(0, Math.min(100, score));
  const label = score >= 85 ? "P0" : score >= 65 ? "P1" : score >= 40 ? "P2" : "P3";
  return { label, score, rationale: rationale.length ? rationale : ["No urgent signals detected."] };
}

function hasAny(text, terms) {
  return terms.some((term) => text.includes(term));
}

File: github_cto/triage.py
from dataclasses import dataclass
from typing import Any


SEVERITY_LABELS = {
    "P0": "Production outage or security incident",
    "P1": "User-facing bug or broken critical workflow",
    "P2": "Important bug or degraded behavior",
    "P3": "Low-risk enhancement or maintenance",
}


@dataclass
class TriageResult:
    severity: str
    score: int
    rationale: list[str]
    recommended_action: str


def triage_issue(issue: dict[str, Any], comments: list[dict[str, Any]] | None = None) -> TriageResult:
    comments = comments or []
    title = (issue.get("title") or "").lower()
    body = (issue.get("body") or "").lower()
    labels = [label.get("name", "").lower() for label in issue.get("labels", [])]
    text = " ".join([title, body, " ".join(labels)])

    score = 25
    rationale: list[str] = []

    critical_terms = ["outage", "down", "p0", "security", "breach", "data loss", "cannot login", "payment"]
    high_terms = ["bug", "crash", "exception", "error", "broken", "regression", "500", "timeout"]
    medium_terms = ["slow", "incorrect", "edge case", "failing test", "flaky"]

    if any(term in text for term in critical_terms):
        score += 45
        rationale.append("Issue contains production-critical or security-sensitive language.")
    if any(term in text for term in high_terms):
        score += 25
        rationale.append("Issue appears to describe a user-facing defect or runtime failure.")
    if any(term in text for term in medium_terms):
        score += 12
        rationale.append("Issue contains reliability or correctness indicators.")
    if "bug" in labels:
        score += 12
        rationale.append("GitHub label marks this as a bug.")
    if "good first issue" in labels or "documentation" in labels:
        score -= 10
        rationale.append("Labels suggest lower operational urgency.")
    if len(comments) >= 5:
        score += 8
        rationale.append("Active discussion suggests meaningful impact or ambiguity.")

    score = max(0, min(score, 100))
    if score >= 85:
        severity = "P0"
        action = "Investigate immediately, create a hotfix branch, and request urgent review."
    elif score >= 65:
        severity = "P1"
        action = "Create a focused fix branch, run tests, and open a PR with incident context."
    elif score >= 40:
        severity = "P2"
        action = "Prepare a normal-priority fix PR and include reproduction notes."
    else:
        severity = "P3"
        action = "Queue for planned maintenance or enhancement work."

    if not rationale:
        rationale.append("No high-risk signals found; defaulting to baseline prioritization.")

    return TriageResult(severity=severity, score=score, rationale=rationale, recommended_action=action)
